src/Build/Definition/Project.cs

```diff
@@ -1117,32 +1129,161 @@ public List<GlobResult> GetAllGlobs(string itemType)
+                return new List<GlobResult>();
+            }
+
+            // scan the project elements in reverse order and build globbing information for each include element
+            // based on the fact that relevant removes for a particular include element (element A) consist of:
+            // - all the removes seen by the next include statement of A's type (element B which appears after A in file order)
+            // - new removes between A and B
+
+            // use immutable lists because there will be a lot of structural sharing between includes which share increasing subsets of corresponding remove elements
+            // item type -> aggregated information about all removes seen so far for that item type
+            var removeElementCache = new Dictionary<string, CumulatedRemoveElementData>(projectItemElements.Count);
+            var globResults = new List<GlobResult>(projectItemElements.Count);
+
+            for (var i = projectItemElements.Count - 1; i >= 0; i--)
+            {
+                var itemElement = projectItemElements[i];
+
+                if (!string.IsNullOrEmpty(itemElement.Include))
+                {
+                    var globResult = BuildGlobResultFromIncludeItem(itemElement, removeElementCache);
+
+                    if (globResult != null)
+                    {
+                        globResults.Add(globResult);
+                    }
+                }
+                else if (!string.IsNullOrEmpty(itemElement.Remove))
+                {
+                    CacheInformationFromRemoveItem(itemElement, removeElementCache);
+                }
+            }
+
+            globResults.TrimExcess();
+
+            return globResults;
         }
 
-        private IEnumerable<GlobResult> GetAllGlobs(ProjectItemElement itemElement)
+        private GlobResult BuildGlobResultFromIncludeItem(ProjectItemElement itemElement, IReadOnlyDictionary<string, CumulatedRemoveElementData> removeElementCache)
         {
-            var includeItemspec = new ItemSpec<ProjectProperty, ProjectItem>(itemElement.Include, _data.Expander, itemElement.IncludeLocation);
-            var excludeItemspec = new ItemSpec<ProjectProperty, ProjectItem>(itemElement.Exclude, _data.Expander, itemElement.ExcludeLocation);
+            var includeItemspec = new EvaluationItemSpec(itemElement.Include, _data.Expander, itemElement.IncludeLocation);
 
-            var excludeSet =
-                new Lazy<IEnumerable<string>>(
-                    () =>
-                        excludeItemspec.Fragments.Where(
-                            // take out item references, we can't reason about them
-                            f => !(f is ItemExpressionFragment<ProjectProperty, ProjectItem>))
-                            .Select(f => f.ItemSpecFragment)
-                            .ToImmutableHashSet());
+            var includeGlobFragments = includeItemspec.Fragments.Where(f => f is GlobFragment).ToImmutableArray();
```

Comment:
I used this as documentation: https://blogs.msdn.microsoft.com/dotnet/2013/06/24/please-welcome-immutablearrayt/
ImmutableArray is backed up by an actual array, and it's a struct. Since there does not seem to be a obvious perf difference from the regular array, I went with the more expressive choice. Is there something I'm missing? :)